url632301a43b_octomarat_Drunkard_tgz-pJ8-ru_drunkard_game_DrunkardTheGameJ8/src/ru/drunkard/field/BasicField.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Override
+    public void setObjectInSector(int x, int y, IFieldObj object) {
+        sectors.get(y).get(x).setObject(object);
+    }
+
+    @Override
+    public void addNewObject(IFieldObj newObject) {
+        lastAdded.add(newObject);
+    }
+
+    @Override
+    public boolean sectorIsEmpty(int x, int y) {
+        return sectors.get(y).get(x).isEmpty();
+    }
+
+    @Override
+    public void sendVisitorToSector(int x, int y, IFieldObj visitor) {
+        sectors.get(y).get(x).acceptVisitor(visitor);
+    }
+
+    @Override
+    public void sendVisitorToSector(int x, int y, GamePrinter visitor) {
+        sectors.get(y).get(x).acceptVisitor(visitor);
+    }
+
+    @Override
+    public List<Point> getFreeNeighbours(Point p, Point special) {
+        List<Point> adjacentVertices = getAllNeighbours(p);
+        List<Point> neighbours = new ArrayList<>();
+        for(Point ap : adjacentVertices) {
+            tryAddFreeNeighbour(ap, special, neighbours);
+        }
+        return neighbours;
+    }
+
+    protected void tryAddValidNeighbour(Point p, List<Point> neighbours) {
+        if(!pointIsOutOfBorders(p)) {
+            neighbours.add(p);
+        }
+    }
+
+    private void tryAddFreeNeighbour(Point p, Point special, List<Point> neighbours) {
+        if(p.equals(special) || sectorIsEmpty(p.x, p.y)) {
+            neighbours.add(p);
+        }
+    }
+
+    @Override
+    public Iterator<Point> getIterator() {
+        return this.new AreaIterator(new Point(0, 0), new Point(width - 1, height - 1));
+    }
+
+    @Override
+    public Iterator<Point> getIterator(Point topleft, Point bottomright) {
+        return this.new AreaIterator(topleft, bottomright);
+    }
+
+    private class AreaIterator implements Iterator<Point> {
+        private final Point startPos;
+        private Point curAbsPos;
+        private final Point endPos;
+
+        public AreaIterator(Point topleft, Point bottomright) {
+            startPos = topleft;
+            curAbsPos = new Point(-1, 0);
+            endPos = bottomright;
+        }
+
+        public boolean hasNext() {
+            return endPos.x != startPos.x + curAbsPos.x ||
+                    endPos.y != startPos.y + curAbsPos.y;
+        }
+
+        public Point next() {
+            if(startPos.x + curAbsPos.x == endPos.x) {
+                curAbsPos.y = (curAbsPos.y + 1) % (endPos.y - startPos.y + 1);
+            }
+            curAbsPos.x = (curAbsPos.x + 1) % (endPos.x - startPos.x + 1);
+            return new Point(startPos.x + curAbsPos.x, startPos.y + curAbsPos.y);
+        }
+
+        //not supported
+        public void remove() {}
+    }
+
+    private void refreshObjectsList() {
+        if(!lastAdded.isEmpty()) {
+            objects.addAll(lastAdded);
+            lastAdded.clear();
+        }
+    }
+
+    @Override
+    public boolean pointIsOutOfBorders(Point p) {
+        return p.x >= width || p.y >= height || p.x < 0 || p.y < 0;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;p&#39; (index: 0 | type: Point). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;tryAddFreeNeighbour&#39;. The same parameter type is validated here for example: url632301a43b_octomarat_Drunkard_tgz-pJ8-ru_drunkard_game_DrunkardTheGameJ8/src/ru/drunkard/fieldobjects/Cop.java:16. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).